Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs

```diff
@@ -96,11 +105,61 @@ private static IEnumerable<MemberResult> GetScopeVariables(ModuleAnalysis analys
             return res;
         }
 
-        private static SymbolKind ToSymbolKind(PythonMemberType memberType) {
-            switch (memberType) {
+        private DocumentSymbol[] ToDocumentSymbols(List<MemberResult> members) {
+            var childMap = new Dictionary<MemberResult, List<MemberResult>>();
+            var topLevel = new List<MemberResult>();
+
+            foreach (var m in members) {
+                var parent = members.FirstOrDefault(x => x.Scope?.Node == m.Scope?.OuterScope?.Node && x.Name == m.Scope?.Name);
+                if (parent != default(MemberResult)) {
+                    if (!childMap.TryGetValue(parent, out var children)) {
+                        childMap[parent] = children = new List<MemberResult>();
+                    }
+                    children.Add(m);
+                } else {
+                    topLevel.Add(m);
+                }
+            }
+
+            var symbols = topLevel
+                .GroupBy(mr => mr.Name)
+                .Select(g => g.First())
+                .Select(m => ToDocumentSymbol(m, childMap))
+                .ToArray();
+
+            return symbols;
+        }
+
+        private DocumentSymbol ToDocumentSymbol(MemberResult m, Dictionary<MemberResult, List<MemberResult>> childMap) {
+            var res = new DocumentSymbol {
+                name = m.Name,
+                detail = m.Name,
+                kind = ToSymbolKind(m),
+                deprecated = false
+            };
+
+            if (childMap.TryGetValue(m, out var children)) {
+                res.children = children.Select(x => ToDocumentSymbol(x, childMap)).ToArray();
+            } else {
+                res.children = new DocumentSymbol[0];
+            }
+
+            var loc = m.Locations.FirstOrDefault(l => !string.IsNullOrEmpty(l.FilePath));
+            if (loc != null) {
+                res.range = new SourceSpan(
+                        new SourceLocation(loc.StartLine, loc.StartColumn),
+                        new SourceLocation(loc.EndLine ?? loc.StartLine, loc.EndColumn ?? loc.StartColumn)
+                    );
+                res.selectionRange = res.range;
+            }
+            return res;
+        }
+
+        private static SymbolKind ToSymbolKind(MemberResult m) {
+            switch (m.MemberType) {
                 case PythonMemberType.Unknown: return SymbolKind.None;
-                case PythonMemberType.Class: return SymbolKind.Class;
-                case PythonMemberType.Instance: return SymbolKind.Object;
+                case PythonMemberType.Class: return m.Scope is FunctionScope ? SymbolKind.Variable : SymbolKind.Class;
```

Comment:
This should be a constant mapping. If nested classes have an incorrect member type, that is probably its own bug.

Alternatively, handle this extra check earlier, so that Scope doesn't have to be passed around here.

Reply:
It a minor nit. In class method first argument comes as class although technically it is an argument in the signature. It looks a bit odd in the outline as a class inside a function. Or maybe its OK?

Reply:
Well, you'd certainly want it to look like a class in this case, so perhaps the answer is to explicitly handle arguments differently?

```python
def make_type():
    class MyClass:
        pass
    return MyClass
```

Reply:
OK